[Doc Improvement][Update preview manifest reference for content add-ins][3679844] (#11252)
MicrosoftDocs/msteams-docs: msteams-platform/resources/schema/manifest-schema-dev-preview.md

To use `extensions.runtimes`, see [create add-in commands](/office/dev/add-ins/develop/create-addin-commands-unified-manifest), [configure the runtime for a task pane](/office/dev/add-ins/develop/create-addin-commands-unified-manifest#configure-the-runtime-for-the-task-pane-command), and [configure the runtime for the function command](/office/dev/add-ins/develop/create-addin-commands-unified-manifest#configure-the-runtime-for-the-function-command).

### extensions.contentRuntimes

The `extensions.contentRuntimes` array configures pages of content that are embedded in an Excel or PowerPoint document. 

|Name| Type| Maximum size | Required | Description|
|---|---|---|---|---|
|`id`| String | 64 characters | ✔️ | Specifies the ID for runtime. Must be unique across all `runtimes` and `contentRuntimes` objects. |
|`code`| Object | | ✔️ | Specifies the location of content for the runtime. |

> |`code.page`| String | 2048 charaters| ✔️ | The full URL of the page that contains the content that is embedded in the document. |

Author: It cannot be used, but it will be seen in Intellisense, so I'll document that it should not be used.

|`requestedWidth`| Number | | | The desired width between 32 and 1000 pixels. If this property isn't used, the Office application determines the width.|
|`requestedHeight`| Number | | | The desired height between 32 and 1000 pixels. If this property isn't used, the Office application determines the height.|

> |`disableSnapshot` | Boolean | | | Specifies whether a snapshot image of your content add-in is saved with the host document. Default value is `false`, so a snapshot is saved. Set to `true` to prevent a snapshot from being saved.|

Reviewer: Hi @Rick-Kirkham,
Please rephrase the sentence to add 'Default value: `false` ' at the end of the sentence for consistency. Also, in other places, we have mentioned 'host' as 'Office application'. We can follow the same here.

|`requirements`| Object | | | Specifies the scopes that must be supported on the Office client in order for the content to be embedded in the document. For more information, see [Specify Office Add-in requirements in the unified manifest for Microsoft 365](/office/dev/add-ins/develop/requirements-property-unified-manifest).|

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Author: ```suggestion
|`requirements.capabilities`| Array | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object |
```

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Author: It cannot be removed because the "contentRuntimes.requirements" property uses the standard data type of "requirements" properties, so "capabilities" will appear in Intellisense even when the grandparent is a content runtimes object.

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Author: > Please check whether this property can be removed since we are adding it under contentRuntimes.

It cannot be removed because the "contentRuntimes.requirements" property uses the standard data type of "requirements" properties, so "capabilities" will appear in Intellisense even when the grandparent is a content runtimes object.

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Reviewer: Even if it appears in intellisense, the developer can decide to add or ignore it. Instead of adding it in the property table, can we add a note that 'requirements.capabilities and requirements.formFactors are not required for contentRuntimes?

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Author: > Even if it appears in intellisense, the developer can decide to add or ignore it. Instead of adding it in the property table, can we add a note that 'requirements.capabilities and requirements.formFactors are not required for contentRuntimes?

We need to say that they should not be used, not that they are "not required". They aren't **required** in ANY scenario. They are always optional. But in most grandparent properties, they will have an effect if they are used. In the case of contentRuntimes, they won't.

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Reviewer: Can we mention the same as note then? My concern of adding it in properties table is, some users will add and check properties in the list intuitively before reading the description. Having the properties not to be added in the table may cause confusion. What are your thoughts?

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Author: > Can we mention the same as note then? My concern of adding it in properties table is, some users will add and check properties in the list intuitively before reading the description. Having the properties not to be added in the table may cause confusion. What are your thoughts?

Please clarify what you mean by "same as note". 

I think users that just check properties in the list will note that "capabilities" and "formFactors" are not checked in the **Required** column. At any rate, we should not design our documentation to accommodate people who don't read the documentation. It is reasonable to assume that people who are looking at these rows of the table are going to read the description. Also, note that people don't use the "requirements" property based only on what they see in this brief reference documentation. They will be educated in the use of that property by reading the conceptual documentation, especially [Specify Office Add-in requirements in the unified manifest for Microsoft 365](https://learn.microsoft.com/en-us/office/dev/add-ins/develop/requirements-property-unified-manifest) which will have a section about "extensions.contentRuntimes.requirements".

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Reviewer: Thanks for adding clarity, @Rick-Kirkham. Give me some time to analyze the user experience of having the properties in the table and having them as a note at the end of the table. Will revert soon.

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Reviewer: Hi @Rick-Kirkham,
As suggested by you, we can add it in the table. Can we edit the description alone since we have never used the word 'grandparent' in our documents?
**Suggestion:** Don't add this property for 'contentRuntimes' object.

> |`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |

Author: > Hi @Rick-Kirkham, As suggested by you, we can add it in the table. Can we edit the description alone since we have never used the word 'grandparent' in our documents? **Suggestion:** Don't add this property for `contentRuntimes` object. (or) This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object.

Yes. I'll use your second suggestion. Thanks!

> |`requirements.capabilities`| Array | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object |

Reviewer: Hi @Rick-Kirkham,
Can you please add the end punctuation to both the sentences?

Author: > Hi @Rick-Kirkham, Can you please add the end punctuation to both the sentences?

@v-ramyas5 Done.

> |`requirements.capabilities`| Array | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object |

Author: ```suggestion
|`requirements.capabilities`| Array | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object. |
```

> |`requirements.scopes`| Array of enums | 1 | | Identifies the scopes in which the add-in can run and defines the Microsoft 365 applications in which the extension can run. For example, `workbook` (Excel). <br>Supported values: `workbook`, `presentation` |

Reviewer: Hi @Rick-Kirkham,
Hope you have confirmed with the feature PM about the supported scopes.

Author: > Hi @Rick-Kirkham, Hope you have confirmed with the feature PM about the supported scopes.

Yes. But note that the whole "requirements" property is going to be removed from the schema for the parent "contentRuntimes". This may not happen in time for public preview, so the property may still be seen in Intellisense. So, I will change this to say that they should not use the "requirements" property when the parent is "contentRuntimes".

> |`requirements.formFactors`| Array of enums | | | Don't use when the grandparent is a content runtime object. |

Author: ```suggestion
|`requirements.formFactors`| Array of enums | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object |
```

> |`requirements.formFactors`| Array of enums | | | Don't use when the grandparent is a content runtime object. |

Author: > Please check whether this property can be removed since we are adding it under contentRuntimes.

It cannot be removed because the "contentRuntimes.requirements" property uses the standard data type of "requirements" properties, so "capabilities" will appear in Intellisense even when the grandparent is a content runtimes object.

> |`requirements.formFactors`| Array of enums | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object |

Author: ```suggestion
|`requirements.formFactors`| Array of enums | | | This property (though present under `requirements` for other objects) must not be added for `contentRuntimes` object. |
```

|`requestedHeight`| Number | | | The desired height between 32 and 1000 pixels. If this property isn't used, the Office application determines the height.|
|`disableSnapshot` | boolean | | | Specifies whether a snapshot image of your content add-in is saved with the host document. Default value is `false`, so a snapshot is saved. Set to `true` to prevent a snapshot from being saved.|
|`requirements`| Object | | | Specifies the scopes, formFactors, and Office JavaScript Library requirement sets that must be supported on the Office client in order for the content to be embedded in the document. For more information, see [Specify Office Add-in requirements in the unified manifest for Microsoft 365](/office/dev/add-ins/develop/requirements-property-unified-manifest).|

> |`requirements.capabilities`| Array | | | Do not use when the grandparent is a content runtime object. |

Reviewer 2: ```suggestion
|`requirements.capabilities`| Array | | | Don't use when the grandparent is a content runtime object. |
```

|`requirements.scopes`| Array of enums | 1 | | Identifies the scopes in which the add-in can run and defines the Microsoft 365 applications in which the extension can run. For example, `workbook` (Excel). <br>Supported values: `workbook`, `presentation` |

> |`requirements.formFactors`| Array of enums | | | Do not use when the grandparent is a content runtime object. |

Reviewer 2: ```suggestion
|`requirements.formFactors`| Array of enums | | | Don't use when the grandparent is a content runtime object. |
```

To use `extensions.runtimes`, see [create add-in commands](/office/dev/add-ins/develop/create-addin-commands-unified-manifest), [configure the runtime for a task pane](/office/dev/add-ins/develop/create-addin-commands-unified-manifest#configure-the-runtime-for-the-task-pane-command), and [configure the runtime for the function command](/office/dev/add-ins/develop/create-addin-commands-unified-manifest#configure-the-runtime-for-the-function-command).

### extensions.contentRuntimes

The `extensions.contentRuntimes` array configures pages of content that are embedded in an Excel or PowerPoint document. 

|Name| Type| Maximum size | Required | Description|
|---|---|---|---|---|
|`id`| String | 64 characters | ✔️ | Specifies the ID for runtime. Must be unique across all `runtimes` and `contentRuntimes` objects. |
|`code`| Object | | ✔️ | Specifies the location of content for the runtime. |
|`code.page`| String | 2048 charaters| ✔️ | The full URL of the page that contains the content that is embedded in the document. |

> |`code.script`| String | | | Don't use when the grandparent is a content runtime object.  |

Reviewer: We avoid the usage of parent or grandparent.

|`id`| String | 64 characters | ✔️ | Specifies the ID for runtime. Must be unique across all `runtimes` and `contentRuntimes` objects. |
|`code`| Object | | ✔️ | Specifies the location of content for the runtime. |
|`code.page`| String | 2048 charaters| ✔️ | The full URL of the page that contains the content that is embedded in the document. |

> Don't use this property under `extensions.contentRuntimes`.

Reviewer: ```Suggestion
|`code.script`| String | | | Don't use this property under `extensions.contentRuntimes`.|
```

To use `extensions.runtimes`, see [create add-in commands](/office/dev/add-ins/develop/create-addin-commands-unified-manifest), [configure the runtime for a task pane](/office/dev/add-ins/develop/create-addin-commands-unified-manifest#configure-the-runtime-for-the-task-pane-command), and [configure the runtime for the function command](/office/dev/add-ins/develop/create-addin-commands-unified-manifest#configure-the-runtime-for-the-function-command).

### extensions.contentRuntimes

The `extensions.contentRuntimes` array configures pages of content that are embedded in an Excel or PowerPoint document. 

|Name| Type| Maximum size | Required | Description|
|---|---|---|---|---|
|`id`| String | 64 characters | ✔️ | Specifies the ID for runtime. Must be unique across all `runtimes` and `contentRuntimes` objects. |
|`code`| Object | | ✔️ | Specifies the location of content for the runtime. |
|`code.page`| String | 2048 charaters| ✔️ | The full URL of the page that contains the content that is embedded in the document. |

> |`code.script`| String | | | Don't use this property under `extensions.contentRuntimes`.|

Reviewer 3: The "Don't use this property" description doesn't really provide a reason not to use it. How about something like:

Not applicable for `extensions.contentRuntimes`.

Reviewer: Hi @Rick-Kirkham,
I like the suggestion made by Erika and incline to change the sentence as suggested. What are your thoughts?

|`requestedWidth`| Number | | | The desired initial width in pixels of the embedded content page. This must be between 32 and 1000. If this property isn't specified, the Office application determines the width.|
|`requestedHeight`| Number | | | The desired initial height in pixels of the embedded content page. This must be between 32 and 1000. If this property isn't specified, the Office application determines the height.|
|`disableSnapshot` | Boolean | | | Specifies whether a snapshot image of your content add-in is saved with the Office application document. Set to `true` to prevent a snapshot from being saved. Default value: `false`|

> |`requirements`| Object | | | Don't use this property under `extensions.contentRuntimes`. |

Reviewer 3: Same for this "Don't use this property" text. (Whatever the final decision is, just noting another instance here to verify we're using consistent language for both).
